Mathlib/Geometry/Euclidean/Incenter.lean

```diff
@@ -0,0 +1,615 @@
+/-
+Copyright (c) 2025 Joseph Myers. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Joseph Myers
+-/
+import Mathlib.Geometry.Euclidean.SignedDist
+import Mathlib.Geometry.Euclidean.Sphere.Basic
+
+/-!
+# Incenters and excenters of simplices.
+
+This file defines the insphere and exspheres of a simplex (tangent to the faces of the simplex),
+and the center and radius of such spheres.
+
+## Main definitions
+
+* `Affine.Simplex.ExcenterExists` says whether an excenter exists with a given set of indices
+  (that determine, up to negating all the signs, which vertices of the simplex lie on the same
+  side of the opposite face as the excenter and which lie on the opposite side of that face).
+* `Affine.Simplex.excenterWeights` are the weights of the excenter with the given set of
+  indices, if it exists, as an affine combination of the vertices.
+* `Affine.Simplex.exsphere` is the exsphere with the given set of indices, if it exists, with
+  shorthands:
+  * `Affine.Simplex.excenter` for the center of this sphere
+  * `Affine.Simplex.exradius` for the radius of this sphere
+* `Affine.Simplex.insphere` is the insphere, with shorthands:
+  * `Affine.Simplex.incenter` for the center of this sphere
+  * `Affine.Simplex.inradius` for the radius of this sphere
+
+## References
+
+* https://en.wikipedia.org/wiki/Incircle_and_excircles
+* https://en.wikipedia.org/wiki/Incenter
+
+-/
+
+
+open EuclideanGeometry
+open scoped Finset RealInnerProductSpace
+
+variable {V P : Type*} [NormedAddCommGroup V] [InnerProductSpace ℝ V] [MetricSpace P]
+variable [NormedAddTorsor V P]
+
+noncomputable section
+
+namespace Affine
+
+namespace Simplex
+
+variable {n : ℕ} [NeZero n] (s : Simplex ℝ P n)
+
+/-- The unnormalized weights of the vertices in an affine combination that gives an excenter with
+signs determined by the given set of indices (for the empty set, this is the incenter; for a
+singleton set, this is the excenter opposite a vertex).  An excenter with those signs exists if
+and only if the sum of these weights is nonzero (so the normalized weights sum to 1). -/
+def excenterWeightsUnnorm (signs : Finset (Fin (n + 1))) (i : Fin (n + 1)) : ℝ :=
```

Comment:
Perhaps slightly nicer to work with as
```suggestion
def excenterWeightsUnnorm (signs : Fin (n + 1) → ℤˣ) (i : Fin (n + 1)) : ℝ :=
```
since then you can skip the `if` below and in the proofs.

Reply:
A clear advantage of writing some kind of set is that there's an obvious canonical way of writing "excenter opposite vertex i" using `{i}`, which becomes longer and less canonical with `Fin (n + 1) → ℤˣ`.

Reply:
I guess I'd argue that's `Pi.mulSingle i (-1)`, but I agree this is less obvious.